File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.rs
impl Solution {
    pub fn my_atoi(s: String) -> i32 {
        let mut index = 0;
        let mut sign = 1;
        let mut result = 0;

        // Ignore leading whitespaces
        while index < s.len() && s.as_bytes()[index] == b' ' {
            index += 1;
        }

        // Determine the sign
        if index < s.len() && (s.as_bytes()[index] == b'+' || s.as_bytes()[index] == b'-') {
            sign = if s.as_bytes()[index] == b'-' { -1 } else { 1 };
            index += 1;
        }

        // Convert digits to integer
        while index < s.len() && s.as_bytes()[index] >= b'0' && s.as_bytes()[index] <= b'9' {
            let digit = (s.as_bytes()[index] - b'0') as i32;
            if result > i32::MAX / 10 || (result == i32::MAX / 10 && digit > 7) {
                return if sign == 1 { i32::MAX } else { i32::MIN };
            }
            result = result * 10 + digit;
            index += 1;
        }

        result * sign
    }
}

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        int index = 0;
        int sign = 1;
        int result = 0;

        // Ignore leading whitespaces
        while (index < s.length() && s.charAt(index) == ' ') {
            index++;
        }

        // Determine the sign
        if (index < s.length() && (s.charAt(index) == '+' || s.charAt(index) == '-')) {
            sign = (s.charAt(index) == '+') ? 1 : -1;
            index++;
        }

        // Convert digits to integer
        while (index < s.length() && Character.isDigit(s.charAt(index))) {
            int digit = s.charAt(index) - '0';
            if (result > Integer.MAX_VALUE / 10 || (result == Integer.MAX_VALUE / 10 && digit > 7)) {
                return (sign == 1) ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }
            result = result * 10 + digit;
            index++;
        }

        return result * sign;
    }
}
